File: atlas_d_engineering_r3/source/verify.mjs
import {chromium} from 'playwright';
import fs from 'node:fs';
import path from 'node:path';
const base=process.env.BASE_URL||'http://127.0.0.1:8755/';
const out=process.env.REVIEW_DIR||'/tmp/atlas-r3-review';fs.mkdirSync(out,{recursive:true});
const browser=await chromium.launch({headless:true,args:['--no-sandbox','--use-gl=angle','--use-angle=swiftshader','--enable-unsafe-swiftshader']});
const results=[];
try{
for(const [label,width,height] of [['desktop',1440,1000],['phone',430,932]]){
 const context=await browser.newContext({viewport:{width,height},deviceScaleFactor:1,hasTouch:label==='phone'});const page=await context.newPage();
 const errors=[],failures=[];page.on('pageerror',e=>errors.push(String(e)));page.on('requestfailed',r=>{if(/ATLAS_D_R3|viewer.js|runtime.js|style.css/.test(r.url()))failures.push(r.url()+':'+r.failure()?.errorText)});
 try{
 const response=await page.goto(base,{waitUntil:'domcontentloaded',timeout:60000});let confirmation=false;
 if(await page.getByRole('button',{name:'Open the page',exact:true}).count()){confirmation=true;await page.getByRole('button',{name:'Open the page',exact:true}).click();}
 await page.waitForFunction(()=>window.ATLAS_D?.ready,{},{timeout:90000});
 const checks=[];for(const progress of [0,.34,.46,.60,.70,.84,1]){await page.evaluate(v=>ATLAS_D.seek(v),progress);checks.push(await page.evaluate(()=>ATLAS_D.diagnostics()))}
 const last=checks.at(-1);if(last.revision!==3||last.scaleTracks!==0||last.hullCount!==6||Math.abs(last.pressureBar-4.0207265)>.0001||last.ratedDepth!==null)throw Error('Physics or continuous-hull validation failed '+JSON.stringify(last));
 if(checks[3].datumDepth!==0||checks.some(c=>c.tenderY!==0||c.waterY!==.012))throw Error('Fixed ocean / staged descent validation failed');
 await page.evaluate(()=>ATLAS_D.reset());await page.locator('#inspect-button').click();await page.locator('#fault').click();
 await page.locator('#timeline').focus();await page.locator('#timeline').press('End');
 const stopped=await page.evaluate(()=>ATLAS_D.diagnostics());if(stopped.progress!==.46||stopped.datumDepth!==0||stopped.pressureHatchesClosed)throw Error('Hatch fault did not inhibit descent '+JSON.stringify(stopped));
 await page.locator('#fault').click();await page.locator('#close-inspector').click();await page.evaluate(()=>ATLAS_D.reset());
 await page.locator('#play').click();await page.waitForTimeout(750);await page.locator('#play').click();if(await page.evaluate(()=>ATLAS_D.progress)<=0)throw Error('Play/pause failed');
 await page.locator('#reverse').click();await page.waitForTimeout(250);await page.evaluate(()=>ATLAS_D.pause());await page.locator('#reset').click();
 await page.waitForTimeout(500);await page.screenshot({path:path.join(out,label+'_surface.png'),timeout:60000});
 await page.evaluate(()=>{ATLAS_D.setXray(true);ATLAS_D.setSystem('pressure')});await page.waitForTimeout(500);await page.screenshot({path:path.join(out,label+'_pressure.png'),timeout:60000});
 await page.evaluate(()=>{ATLAS_D.seek(.84);ATLAS_D.setSystem('ballast')});await page.waitForTimeout(300);await page.screenshot({path:path.join(out,label+'_ballast.png'),timeout:60000});
 await page.evaluate(()=>{ATLAS_D.seek(1);ATLAS_D.setSystem('wet')});await page.waitForTimeout(300);await page.screenshot({path:path.join(out,label+'_wet.png'),timeout:60000});
 await page.evaluate(()=>{ATLAS_D.setXray(false);ATLAS_D.setView('hero')});await page.waitForTimeout(300);await page.screenshot({path:path.join(out,label+'_submerged.png'),timeout:60000});
 await page.evaluate(()=>{ATLAS_D.emergency();ATLAS_D.recovery.started-=19000});await page.waitForFunction(()=>Math.abs(ATLAS_D.depth)<.01);const recovery=await page.evaluate(()=>ATLAS_D.diagnostics());if(!recovery.pressureHatchesClosed)throw Error('Recovery reopened pressure boundary');
 await page.locator('#reset').click();
 results.push({label,width,height,url:page.url(),httpStatus:response.status(),hostConfirmationRequired:confirmation,physicsChecks:checks,hatchFault:stopped,recovery,errors,assetFailures:failures,passed:errors.length===0&&failures.length===0});
 }catch(e){fs.writeFileSync(path.join(out,label+'_error.json'),JSON.stringify({error:String(e),errors,failures,body:await page.locator('body').innerText().catch(()=>''),runtime:await page.evaluate(()=>window.ATLAS_D?.diagnostics?.()||window.ATLAS_D?.error).catch(()=>null)},null,2));await page.screenshot({path:path.join(out,label+'_failure.png'),timeout:15000}).catch(()=>{});throw e}finally{await context.close()}
}
if(results.some(r=>!r.passed))throw Error('Browser errors');
}finally{fs.writeFileSync(path.join(out,'validation.json'),JSON.stringify({url:base,passed:results.length===2&&results.every(r=>r.passed),physicalIPhoneTest:false,engineeringValidated:false,tests:results},null,2));await browser.close();}

File: atlas_d_integrated_r5/source/verify.py
"""Browser checks of the shipped geometry/controls, not a safety validation."""
from pathlib import Path
import os,json,time
from playwright.sync_api import sync_playwright
URL=os.getenv('BASE_URL','http://127.0.0.1:8785/')
OUT=Path(os.getenv('REVIEW_DIR','/tmp/atlas5/review'));OUT.mkdir(parents=True,exist_ok=True)
report={'url':URL,'physical_iPhone_test':False,'engineering_validation':False,'tests':[]}
with sync_playwright() as p:
 browser=p.chromium.launch(headless=True,args=['--no-sandbox','--use-gl=angle','--use-angle=swiftshader','--enable-unsafe-swiftshader'])
 for label,w,h in [('desktop',1500,960),('phone',430,932)]:
  ctx=browser.new_context(viewport={'width':w,'height':h},device_scale_factor=1)
  page=ctx.new_page();errors=[];failures=[]
  page.on('pageerror',lambda e:errors.append(str(e)))
  page.on('requestfailed',lambda req:failures.append({'url':req.url,'failure':req.failure}) if any(x in req.url for x in ['ATLAS_D_R5.glb','runtime.js','viewer.js','style.css']) else None)
  start=time.monotonic();res=page.goto(URL,wait_until='domcontentloaded',timeout=90000)
  confirm=False
  if 'One more step' in page.locator('body').inner_text():
   confirm=True;page.get_by_role('button',name='Open the page',exact=True).click()
  page.wait_for_function('window.ATLAS_R5?.ready',timeout=90000)
  page.wait_for_function('window.ATLAS_R5.report?.revision===5',timeout=30000)
  states=[]
  for name,v,t in [('surface','hero',0),('halfway','hero',.3),('stowed','hero',.62),('profile','profile',0),('aft','aft',0),('canopy','canopy',.29),('mast','mast',.2),('terraces','terraces',.35),('submerged','hero',1)]:
   page.evaluate('([v,p])=>{ATLAS_R5.pose(p);ATLAS_R5.setView(v)}',[v,t]);page.wait_for_timeout(550)
   states.append(page.evaluate('ATLAS_R5.inspect()'))
   page.screenshot(path=str(OUT/f'{label}_{name}.png'),timeout=60000)
  assert abs(states[2]['canopy'][0]+5.2)<.002
  assert abs(states[2]['mast'][1]-9.74)<.002
  assert abs(states[2]['depth'])<.0001
  assert states[-1]['depth']==26
  page.locator('#reset').click();page.locator('#compare').click();assert abs(page.evaluate('ATLAS_R5.progress')-.62)<.001
  page.locator('#compare').click();assert page.evaluate('ATLAS_R5.progress')==0
  page.locator('#core').click();page.wait_for_timeout(500);page.screenshot(path=str(OUT/f'{label}_core.png'))
  page.locator('#core').click();page.locator('#bays').click();page.wait_for_timeout(500);page.screenshot(path=str(OUT/f'{label}_bays.png'))
  page.locator('#bays').click();page.locator('#reset').click()
  page.evaluate('ATLAS_R5.pose(.61)');page.locator('#convert').click();page.wait_for_function('!ATLAS_R5.playing',timeout=7000)
  assert abs(page.evaluate('ATLAS_R5.progress')-.62)<.001
  assert not errors,errors
  assert not failures,failures
  item={'label':label,'http_status':res.status,'host_confirmation_required':confirm,'elapsed_test_seconds':round(time.monotonic()-start,2),'poses':states,'errors':errors,'asset_failures':failures,'controls':['convert stops above water','before/stowed same camera','timeline','pressure core','stowage inspection','reset'],'passed':True}
  report['tests'].append(item);ctx.close()
 browser.close()
report['passed']=True;(OUT/'validation.json').write_text(json.dumps(report,indent=2));print(json.dumps(report,indent=2))
